package-scripts/datadog-signing-keys/postinst

```diff
@@ -23,12 +25,22 @@ USE_TRUSTED_GPG_D=true
 DISTRIBUTION_VERSION=$(. /etc/os-release && echo "${VERSION_ID}" | tr "." " " | awk '{ print $1 }')
 
 if { [ "${DISTRIBUTION}" = "Debian" ] && [ "${DISTRIBUTION_VERSION}" -ge "9" ]; } || { [ "${DISTRIBUTION}" = "Ubuntu" ] && [ "${DISTRIBUTION_VERSION}" -ge "16" ]; } ; then
+    DD_LIST_SIGNED_BY=false
+    OPW_LIST_SIGNED_BY=false
     # We also want to create the /etc/apt/trusted.gpg.d keyring if the system
     # is not set to use the /usr/share/keyrings/ keyring through signed-by.
     # This might create the /etc/apt/trusted.gpg.d keyring unnecessarily on
     # some systems, but it's the only way to be at least somewhat certain
     # that anyone who installed the package is ok.
-    if grep "signed-by=${USR_SHARE_KEYRING}" /etc/apt/sources.list.d/datadog.list 2>/dev/null 1>/dev/null; then
+    if [ ! -e "${DD_SOURCE_LIST}" ] || grep "signed-by=${USR_SHARE_KEYRING}" "${DD_SOURCE_LIST}" 2>/dev/null 1>/dev/null; then
+        DD_LIST_SIGNED_BY=true
+    fi
+
+    if [ ! -e "${OPW_SOURCE_LIST}" ] || grep "signed-by=${USR_SHARE_KEYRING}" "${OPW_SOURCE_LIST}" 2>/dev/null 1>/dev/null; then
+        OPW_LIST_SIGNED_BY=true
+    fi
+
+    if [ "${DD_LIST_SIGNED_BY}" = "true" ] && [ "${OPW_LIST_SIGNED_BY}" = "true" ]; then
         USE_TRUSTED_GPG_D=false
     fi
```

Comment:
Nit: It took me a bit to understand the logic here, because the value of `DD_LIST_SIGNED_BY` represents "(the list file contains the signed-by mention) or (the list file doesn't exist)", which isn't obvious.

Reversing the variable (to `DD_LIST_NO_SIGNED_BY`) would also be convoluted to read, so I don't know what's the best here.

Also, I believe this changes the behavior when no source files are there. Previously, this would result in `USE_TRUSTED_GPG_D=true`, now you'll get `USE_TRUSTED_GPG_D=false` (because both `DD_LIST_SIGNED_BY` and `OPW_LIST_SIGNED_BY` are true if the files aren't there). In theory, this case shouldn't happen, so I don't know if we really need to fix that behavior change.

Reply:
Yes, you're right that the variable naming isn't 100 % ideal. I don't have a good idea on how to rename these, but can certainly treat it as a thing to improve in the future.

I am aware of the behavior change and it is intentional, but now that you mention it I'm unsure about it TBH. The cases where this might happen are:

* User installed all our packages from downloaded *.deb files (without setting up source lists) - if they keep doing this, they don't need the trusted.gpg.d file (but there's no way to tell if they *will* need it).
* User named our source.list file differently and didn't use signed-by - in this case it would be useful to create the trusted.gpg.d file anyway.

I'm starting to think that we actually want to cover this case (worst case we create the keyring file when it's not necessary on some non-standard setups, which honestly shouldn't hurt). I'll change this.

Reply:
Thinking about this further, this would lead to the logic conclusion that if *any* of the sources.list files isn't present, we should create the trusted.gpg.d file (because what if it was named differently), which at least initially would be vast majority of the cases, so in the end I think actually this is a desired change in the behavior and I'd propose leaving it as it is. WDYT?

Reply:
Obviously this is something to document for users for whenever we do the next key rotation.

Reply:
I did more research and came up with this:

* While I originally thought that `apt-key` removal is a security-directed step meant to prevent adding GPG keys that would apply to all directories, it seems it's not. At least originally, it was meant to prevent operations from depending on gnupg: https://github.com/Debian/apt/commit/ee284d5917d09649b68ff1632d44e892f290c52f.
* The manpage for `apt-key` discusses placing keyrings in `/etc/apt/trusted.gpg.d/` as an option going forward, but recommends placing them elsewhere and use the `signed-by` option.
* The [Debian Wiki page about configuring third party repositories](https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_Key_distribution) states `The key MUST NOT be placed in /etc/apt/trusted.gpg.d or loaded by apt-key add.`, leaving usage of the `signed-by` option as the only alternative.

Where that leaves us with this PR:
* For users that do use our official installation methods/docs, the keyring in `/etc/apt/trusted.gpg.d` will never be created by this package, as we will name the files as expected by this script.
* For users who do certain different things (see [my previous comment](https://github.com/DataDog/datadog-signing-keys/pull/27#discussion_r1028327985)), we will not create the keyring in `/etc/apt/trusted.gpg.d`, possibly causing problems if they didn't use `signed-by` in their custom-named source list files.
  * The question is how often do users use custom-named source list files that point to our repositories and would thus need this keyring in `/etc/apt/trusted.gpg.d`. Unfortunately we don't really have any data about how often this happens.

The original decision when making this script was "if we don't know that the user is safe (== using `datadog.list` which has the proper `signed-by` option inside), it's better to create the `/etc/apt/trusted.gpg.d` keyring to be safe". If we wanted to apply that logic to both of our source list files now, we would have to create the keyring always, unless both of the repofiles are present at the same time (this is likely going to be very rare initially).

It *is* however true, that it's more likely (my opinion) that the customer is using official installation methods either for both of our source.list files or for none of them. That, in my mind, seems to suggest that it is an acceptable workaround to create the `/etc/apt/trusted.gpg.d` keyring if *both* of the files are missing (which is what translates to "if we don't know" case from the previous paragraph), but not if only one of them is. I'll add a special case with a clear comment to handle this case.

Reply:
This logic looks good to me 👍.

So the end state, as I understand it, is: "if at least one known source file is detected and contains the `signed-by` option, then we don't add the keyring in `/etc/apt/trusted.gpg.d`"?

Reply:
Yes, that is exactly right.

Reply:
Update: from the other thread, the correct way to understand this is: "if at least one known source file is detected, _and all detected source files_ contain the signed-by option, then we don't add the keyring in `/etc/apt/trusted.gpg.d`"